src/community/oseo/oseo-service/src/main/java/org/geoserver/opensearch/eo/response/TemplatesProcessor.java

```diff
@@ -227,4 +518,299 @@ private Geometry toGeometry(Object argument) throws TemplateModelException {
+                        request.getBaseUrl(),
+                        "oseo/search/description",
+                        kvp,
+                        URLMangler.URLType.SERVICE);
+        model.put("search", encodeLink("search", href, DescriptionResponse.OS_DESCRIPTION_MIME));
+    }
+
+    private String buildCollectionIdentifierLink(Object identifier, SearchRequest request) {
+        String baseURL = request.getBaseUrl();
+        Map<String, String> kvp = new LinkedHashMap<>();
+        kvp.put("uid", String.valueOf(identifier));
+        kvp.put("httpAccept", AtomSearchResponse.MIME);
+        return ResponseUtils.buildURL(baseURL, "oseo/search", kvp, URLMangler.URLType.SERVICE);
+    }
+
+    private String buildProductIdentifierLink(Object identifier, SearchRequest request) {
+        String baseURL = request.getBaseUrl();
+        Map<String, String> kvp = new LinkedHashMap<>();
+        kvp.put("parentId", request.getParentIdentifier());
+        kvp.put("uid", String.valueOf(identifier));
+        kvp.put("httpAccept", AtomSearchResponse.MIME);
+        return ResponseUtils.buildURL(baseURL, "oseo/search", kvp, URLMangler.URLType.SERVICE);
+    }
+
+    private String buildQuicklookLink(String identifier, SearchRequest request) {
+        String baseURL = request.getBaseUrl();
+        Map<String, String> kvp = new LinkedHashMap<>();
+        kvp.put("parentId", request.getParentIdentifier());
+        kvp.put("uid", String.valueOf(identifier));
+        return ResponseUtils.buildURL(baseURL, "oseo/quicklook", kvp, URLMangler.URLType.SERVICE);
+    }
+
+    private String buildMetadataLink(
+            String parentIdentifier, Object identifier, String mimeType, SearchRequest request) {
+        String baseURL = request.getBaseUrl();
+        Map<String, String> kvp = new LinkedHashMap<>();
+        if (parentIdentifier != null) {
+            kvp.put("parentId", parentIdentifier);
+        }
+        kvp.put("uid", String.valueOf(identifier));
+        if (mimeType != null) {
+            kvp.put("httpAccept", mimeType);
+        }
+        return ResponseUtils.buildURL(baseURL, "oseo/metadata", kvp, URLMangler.URLType.SERVICE);
+    }
+
+    private String buildOsddLink(String parentIdentifier, SearchRequest request) {
+        String baseURL = request.getBaseUrl();
+        Map<String, String> kvp = new LinkedHashMap<>();
+        if (parentIdentifier != null) {
+            kvp.put("parentId", parentIdentifier);
+        }
+        return ResponseUtils.buildURL(baseURL, "oseo/description", kvp, URLMangler.URLType.SERVICE);
+    }
+
+    private String encodeLink(String rel, String builder, String type) {
+        return "href=\""
+                + builder.replaceAll("&", "&amp;")
+                + "\" rel=\""
+                + rel
+                + "\" type=\""
+                + type
+                + "\"";
+    }
+
+    private String encodeLink(String rel, String builder, String type, String title) {
```

Comment:
All these bits of XML (this method and the ones similar to it) should be part of the freemarker template. The whole idea of a template is to have flexibility to customize the output, having hard-coded sequences of attributes does not help.
Please move them to the freemarker template. 

Escaping of & into &amp; should probably be done by Freemarker itself, putting it in XML mode, see: 
https://freemarker.apache.org/docs/dgui_misc_autoescaping.html

Also link building should not be hard-coded, but part of the template, so that the links can be modified.
One can have a support function to build a URL out of variables if needed (see the support available in the OGC modules) but links should be built, as much as possible, directly from the Feature objects, in the FreemarkerTemplate.

The whole point of this exercise is not to build the same output using a different approach, but allow full customization of the output, it does not have to have bits built in Java, as they cannot be customized.

Btw, did you check that the previous and current output,  using the default templates, are the same?

Reply:
I have changed java code to bits as much as possible. The remaining ones are using Java Optionals, DateTime and having complex logic that cannot be transformed to freemarker.

I have used dif checkers for the below tests.
SearchTest.testAllCollection
SearchTest.testAllSentinel2Products